### PR TITLE
Extend contact data with force, normal, and penetration depth

### PR DESCRIPTION
**Pull request by diegoferigo**
This PR provides the support of exporting optional contact data that the physics engine might provide. In particular:

1. Contact force
1. Contact normal
1. Penetration depth

This PR also extends the DART implementation to fill the new data structure containing the additional contact data.

I'm not yet very familiar with all the templates of ign-physics. This is the first working implementation I managed to obtain, I'm not sure if there's a better way to handle the data exchanged between the public APIs and the implementation method.

In the current form, ABI is likely broken. Given the field `CompositeData extraData`, I'm not sure if there's any other way to get that data without the need to define a new public struct.

Closes https://github.com/ignitionrobotics/ign-physics/issues/24.

### CODE REVIEW

**Comment by diegoferigo** [comment]
cc @azeey @scpeters

**Comment by diegoferigo** [comment]
Related issue for visualizing contacts https://github.com/ignitionrobotics/ign-gazebo/issues/112

File: dartsim/src/SimulationFeatures.cc [hunk 1]
```diff
@@ -82,8 +83,14 @@ SimulationFeatures::GetContactsFromLastStep(const Identity &_worldID) const
       std::size_t shape2ID =
           this->shapes.IdentityOf(dtShapeFrame2->asShapeNode());
 
-      // TODO(addisu) Add normal, depth and wrench to extraData.
       CompositeData extraData;
+
+      // Add normal, depth and wrench to extraData.
+      auto& extraContactData = extraData.Get<ExtraContactData>();
```
Comment (azeey) [inline dartsim/src/SimulationFeatures.cc:89]: nit:
```suggestion
      auto &extraContactData = extraData.Get<ExtraContactData>();
```

File: include/ignition/physics/GetContacts.hh [hunk 2]
```diff
@@ -33,12 +33,30 @@ namespace physics
 class IGNITION_PHYSICS_VISIBLE GetContactsFromLastStepFeature
     : public virtual FeatureWithRequirements<ForwardStep>
 {
+  public: template <typename PolicyT>
+  struct ExtraContactDataT
+  {
+     using Scalar = typename PolicyT::Scalar;
+     using VectorType = typename FromPolicy<PolicyT>::template Use<Vector>;
```
Comment (azeey) [inline include/ignition/physics/GetContacts.hh:40]: nit:
```suggestion
    using Scalar = typename PolicyT::Scalar;
    using VectorType = typename FromPolicy<PolicyT>::template Use<Vector>;
```

File: include/ignition/physics/detail/GetContacts.hh [hunk 3]
```diff
@@ -54,6 +54,14 @@ auto GetContactsFromLastStepFeature::World<
...
     auto &contactOutput = output.emplace_back();
     contactOutput.template Get<ContactPoint>() = std::move(contactPoint);
+
+    auto *extraContactData =
+        contact.extraData.template Query<ExtraContactData>();
+
+    if (extraContactData) {
+        contactOutput.template Get<ExtraContactData>() =
+            std::move(*extraContactData);
+    }
```
Comment (azeey) [inline include/ignition/physics/detail/GetContacts.hh:64]: nit: curly brace on new line
```suggestion
    if (extraContactData)
    {
      contactOutput.template Get<ExtraContactData>() =
          std::move(*extraContactData);
    }

```

File: include/ignition/physics/GetContacts.hh [hunk 4]
```diff
@@ -33,12 +33,30 @@ namespace physics
...
+
+    /// \brief The contact force from body acting on the first body
```
Comment (azeey) [inline include/ignition/physics/GetContacts.hh:42]: "The contact force ~from body~ acting on the first body"?

**Comment by azeey** [comment]
Friendly ping @diegoferigo. If it's okay with you, I can commit those suggestions and merge this PR.

**Comment by diegoferigo** [comment]
Hi @azeey, thanks for the review and sorry for the delay. The requested changes are trivial and I just updated this PR. 

I was trying to find some time to address the tests request. As far as I saw, in the past a common pattern to test physics features was opening a corresponding PR in ign-gazebo and add a test there. In this case it would also require updating the Physics system to extract the new contact data. I haven't yet found enough time to start this process.

**Comment by diegoferigo** [comment]
Thanks for the link, I'll have a look to the existing test and check how to update it with the extra contact data.

**Comment by diegoferigo** [comment]
@azeey in 03ef01a I extended the existing test and added a check of the contact normal and contact force.

I haven't yet figured out how to extract the gravity vector from the `world` object, and the mass from each `sphere->GetLink(•)`. Do you have any suggestion? If it's easier editing directly the commit instead of writing how to do it, go ahead :)

**Comment by azeey** [comment]
@diegoferigo, I'm okay with hardcoding the mass values since we don't have a convenient way to retrieve them right now. Looks like I don't have permission to push to your branch directly. Here's the patch I was going to push?

```diff
diff --git a/dartsim/src/SimulationFeatures_TEST.cc b/dartsim/src/SimulationFeatures_TEST.cc
index 478decb..265a7b7 100644
--- a/dartsim/src/SimulationFeatures_TEST.cc
+++ b/dartsim/src/SimulationFeatures_TEST.cc
@@ -264,6 +264,16 @@ TEST_P(SimulationFeatures_TEST, RetrieveContacts)
       {sphere->GetLink(3)->GetShape(0), {1.0, 1.0, 0.0}},
     };
 
+    const double gravity = 9.8;
+    std::map<TestShapePtr, double> forceExpectations
+    {
+      // Contact force expectations are: link mass * gravity.
+      {sphere->GetLink(0)->GetShape(0), 0.1 * gravity},
+      {sphere->GetLink(1)->GetShape(0), 1.0 * gravity},
+      {sphere->GetLink(2)->GetShape(0), 2.0 * gravity},
+      {sphere->GetLink(3)->GetShape(0), 3.0 * gravity},
+    };
+
     for (auto &contact : contacts)
     {
       const auto &contactPoint = contact.Get<ContactPoint>();
@@ -277,17 +287,17 @@ TEST_P(SimulationFeatures_TEST, RetrieveContacts)
       EXPECT_NE(contactPoint.collision1, contactPoint.collision2);
 
       Eigen::Vector3d expectedContactPos = Eigen::Vector3d::Zero();
-      // One of the two collisions is the ground plane and the other is the
-      // collision we're interested in.
-      try
-      {
-        expectedContactPos = expectations.at(contactPoint.collision1);
-      }
-      catch (...)
+
+      // The test expectations are all on the collision that is not the ground
+      // plane.
+      auto testCollision = contactPoint.collision1;
+      if (testCollision == groundPlaneCollision)
       {
-        expectedContactPos = expectations.at(contactPoint.collision2);
+        testCollision = contactPoint.collision2;
       }
 
+      expectedContactPos = expectations.at(testCollision);
+
       EXPECT_TRUE(ignition::physics::test::Equal(expectedContactPos,
                                                  contactPoint.point, 1e-6));
 
@@ -304,9 +314,8 @@ TEST_P(SimulationFeatures_TEST, RetrieveContacts)
       // the the weight of the sphere times the gravitational acceleration
       EXPECT_NEAR(extraContactData->force[0], 0.0, 1e-3);
       EXPECT_NEAR(extraContactData->force[1], 0.0, 1e-3);
-      // mass = ?
-      // gravity = ?
-      // EXPECT_NEAR(extraContactData->force[2], mass * gravity, 1e-3);
+      EXPECT_NEAR(extraContactData->force[2],
+                  forceExpectations.at(testCollision), 1e-3);
     }
   }
 }
diff --git a/include/ignition/physics/detail/GetContacts.hh b/include/ignition/physics/detail/GetContacts.hh
index 0de81b2..80e7584 100644
--- a/include/ignition/physics/detail/GetContacts.hh
+++ b/include/ignition/physics/detail/GetContacts.hh
@@ -63,7 +63,6 @@ auto GetContactsFromLastStepFeature::World<
       contactOutput.template Get<ExtraContactData>() =
           std::move(*extraContactData);
     }
-
   }
   return output;
 }
```

**Comment by diegoferigo** [comment]
> Looks like I don't have permission to push to your branch directly.

Mmh strange. 

> I'm okay with hardcoding the mass values since we don't have a convenient way to retrieve them right now.

Awesome, thanks a lot for the patch @azeey! Branch updated.

**Comment by azeey** [comment]
@osrf-jenkins run tests

**Comment by azeey** [comment]
Looks like a test is failing. I'm looking into it.

**Comment by diegoferigo** [comment]
CI went green, thanks a lot @azeey for the fix in #77 :rocket:

**Pull request merged**